src/core/qfieldcloudprojectsmodel.cpp

```diff
@@ -368,229 +379,484 @@ void QFieldCloudProjectsModel::downloadProject( const QString &projectId, bool o
+    if ( index < 0 || index >= mCloudProjects.size() )
+      return;
 
-    downloadStatusReply->deleteLater();
-    mCloudProjects[index].apiNetworkReply = nullptr;
+    QNetworkReply *rawReply = reply->reply();
 
     if ( rawReply->error() != QNetworkReply::NoError )
     {
-      QgsLogger::debug( QStringLiteral( "Packaging status request for \"%1\" finished with error: %2" ).arg( projectId, rawReply->error() ) );
-      projectDownloadFinishedWithError( projectId, QFieldCloudConnection::errorString( rawReply ) );
+      mCloudProjects[index].jobs[jobType].status = JobFailedStatus;
+      emit projectJobFinished( projectId, jobType, QFieldCloudConnection::errorString( rawReply ) );
       return;
     }
 
     const QJsonObject payload = QJsonDocument::fromJson( rawReply->readAll() ).object();
-    Q_ASSERT( !payload.isEmpty() );
-
-    mCloudProjects[index].packagingStatus = packagingStatus( payload.value( QStringLiteral( "status" ) ).toString() );
+    const QString jobStatusString = payload.value( QStringLiteral( "status" ) ).toString();
 
-    QgsLogger::debug( QStringLiteral( "Packaging request for \"%1\" finished with no error and status \"%2\"" )
-                        .arg( projectId )
-                        .arg( QVariant::fromValue( mCloudProjects[index].packagingStatus ).toString() ) );
-
-    switch ( mCloudProjects[index].packagingStatus )
+    if ( jobStatusString.isEmpty() )
     {
-      case PackagingUnstartedStatus:
-        // download export job should be already started!!!
-        Q_ASSERT( 0 );
-        FALLTHROUGH
+      mCloudProjects[index].jobs[jobType].status = JobFailedStatus;
+      emit projectJobFinished( projectId, jobType, tr( "job(%1) status response does not contain all the expected keys: status(string)" ).arg( jobStatusString ) );
+      return;
+    }
 
-      case PackagingAbortStatus:
-        return;
+    mCloudProjects[index].jobs[jobType].status = jobStatus( jobStatusString );
 
-      case PackagingPendingStatus:
-      case PackagingBusyStatus:
+    switch ( mCloudProjects[index].jobs[jobType].status )
+    {
+      case JobPendingStatus:
+      case JobQueuedStatus:
+      case JobStartedStatus:
+      case JobStoppedStatus:
         // infinite retry, there should be one day, when we can get the status!
         QTimer::singleShot( sDelayBeforeStatusRetry, [=]() {
-          projectGetPackagingStatus( projectId );
+          projectGetJobStatus( projectId, jobType );
         } );
         break;
 
-      case PackagingErrorStatus:
+      case JobFailedStatus:
+        emit projectJobFinished( projectId, jobType, tr( "Job(%1) finished execution with failure status." ) );
+        return;
+      case JobFinishedStatus:
+        emit projectJobFinished( projectId, jobType );
+        return;
+    }
+  } );
+}
+
+void QFieldCloudProjectsModel::projectPackageAndDownload( const QString &projectId )
+{
+  if ( !mCloudConnection )
+    return;
+
+  int index = findProject( projectId );
```

Comment:
Could `findProject` return a pointer to a project, `nullptr` if not found?
That would reduce index range validity checks as done on the line below, lookups (for each `mCloudProjects[index]` and would also solve the current strange VS compiler errors.

Reply:
Makes sense, will try.

Reply:
Check #2369. Will rebase once gets merged.